### ISSUE TITLE
Help with Llama 3.2 vision interface straight through Python

### ISSUE THREAD

**Issue by dfloreaa** [pytorch/torchtune#2112]
For a research project I am trying to access both the textual and image embeddings used for aligning the encoders using CLIP for the Llama 3.2 model. Currently in your documentation there is no clear description of the complete pipeline that inference using the Llama 3.2 model looks like and I would love it if you could help me out with that.

I am currently making an encoder using the same arguments as [llama3_2_vision_11b](https://pytorch.org/torchtune/0.3/_modules/torchtune/models/llama3_2_vision/_model_builders.html#llama3_2_vision_11b) via:
```python
encoder = llama3_2_vision_encoder(
        patch_size=14,
        num_heads=16,
        clip_embed_dim=1280,
        clip_num_layers=32,
        clip_hidden_states=[3, 7, 15, 23, 30],
        decoder_embed_dim=4096,
        num_layers_projection=8,
        tile_size=img_size,
        max_num_tiles=4,
        in_channels=3,
    )

max_seq_len=131_072

vocab_size=128_256
num_special_tokens=8
embed_dim=4096
```

I also managed to extract the weights from the pretrained model and am succesfully loading them using
```python
model_state_dict = torch.load('<PATH>', mmap=True, weights_only=True)
encoder.load_state_dict(model_state_dict)
```

For a random image contained inside of `img` resized to 620x620 I am able to obtain the image embeddings via
```python
# Parameters
batch_size = 1  # Example batch size
num_images = 1  # Example number of images
tiles = 1       # No tiling, so t = 1
channels, height, width = img.shape

# Reshape to [b, i, t, c, w, h]
reshaped = img.unsqueeze(0).unsqueeze(0).unsqueeze(2)  # Add dimensions for b, i, and t
reshaped = reshaped.view(batch_size, num_images, tiles, channels, height, width) / 255.0

img_embeddings = encoder(reshaped)
```
These embeddings have shape `[1, 1601, 4096]`

As for the text embeddings, I am tokenizing my example text via:
```python
tokenizer = llama3_tokenizer(
            "models/Llama-3.2-11B-Vision-Instruct/original/tokenizer.model",
            special_tokens_path="models/Llama-3.2-11B-Vision-Instruct/tokenizer.json",
            max_seq_len=max_seq_len
        )

tokenized_input = torch.IntTensor(tokenizer.encode(text="Hello", add_bos=True, add_eos=True)).unsqueeze(0)
tok_embeddings = FusionEmbedding(vocab_size, num_special_tokens, embed_dim)
input_token_embeddings = tok_embeddings(tokenized_input)
```

These embeddings have shape `[1, 3, 4096]`

According to the CLIP publication, both of these tensors should be the same size and dimensions in order to compute the cosine similarity between the two, therefore, something has to be going on in the way I compute the embeddings that is not allowing me to project them into the same space.

Could I please get some assistance with this? I am looking for a way to access the CLIP embeddings for the text and the image separately, I am also guessing these come from pretrained models with weights that are included with the Llama 3.2 11B Vision weights from Meta.

**Comment by felipemello1** [comment]
@pbontrager , would you mind taking a look?

**Comment by dfloreaa** [comment]
Apparently, the `clip_vision_encoder` created inside of the `llama3_2_vision_encoder` needs to be created with `output_cls_projection=True` (it's set to `output_cls_projection=False` and cannot be changed when creating the encoder).

If this flag is set to true, the CLIP encoder will create a projection layer via `CLSProjection(embed_dim=embed_dim, cls_output_dim=cls_output_dim)`, but apparently, no weights for this projection layer are provided when adapting the weights from Meta using `meta_to_tune()`. Is there a way to acces these weights? If not, there is no way to access the `[1, 4096]` vector containing the image's embeddings.

I have not taken a proper look into it, but I guess there is a similar projection head for the text embeddings (or should I just take the embeddings corresponding to the EOT token?)

Your help would be very much appretiated

**Comment by dfloreaa** [comment]
From what I am gettings, Llama 3.2 took a pretrained CLIP image encoder and put a projection layer on top of it to align the encoder's output to the Llama Tokenizer + FusionEmbeddings space and further finetuned the CLIP encoder. Therefore, it may not possess the ability to properly embed the image to a joint space with the text on it's CLS token embeddings. 

As for my application, I am trying to add a few adapter modules inbetween the image and text encoders and the decoder input. Similarly to CLIP, I want to use contrastive learning to properly align the output representations of the image and the text before they enter the decoder.

If I were to follow the CLIP approach, aligning the image embeddings from the CLS token and the text embeddings for the EOT token would benefit from using the pretrained open CLIP weights, whereas trying to do so with the Llama weights would not yield such advantage.

My question then is, how was Llama 3.2 further trained to align image and textual inputs? You had to somehow use a method similar to CLIP in order to make both encoders project into the same space (actually, adjusting the image encoder to project into the Llama3 text embedding space). How was this performed? Did you also just do contrastive learning on the CLS and EOT tokens? Could I just do the same on my project?

Thank you very much for your extensive reply to my previous questions, I am currently deciding whether to stick to Llama3.2 or find another LLM that works with CLIP embeddings directly.

**Comment by pbontrager** [comment]
> My question then is, how was Llama 3.2 further trained to align image and textual inputs? You had to somehow use a method similar to CLIP in order to make both encoders project into the same space (actually, adjusting the image encoder to project into the Llama3 text embedding space). How was this performed?

The CLIP pre-training loss is only useful for having good initial embeddings. But once the model is merged as a VLM architecture then the only loss is next token prediction. The gradients flow back to the CLIP model and update it such that the text model can produce better text. It doesn't really matter whether the image embeddings are in a joint space with text embeddings as long as the cross attention layers in the model work.

> Did you also just do contrastive learning on the CLS and EOT tokens? Could I just do the same on my project?

You could do this to try and recreate a CLIP style model, but for a VLLM the only loss is the text decoder prediction.

While the models aren't strictly trained to have a joint embedding, you could they likely do maintain some of those properties since they cross attend with text tokens. Since Llama 3.2 is a deepfusion model, the encoder inputs are fed into cross attention layers in intermediate layers in the model which is not really what you're looking for. You probably want an EarlyFusion model (Like PaliGemma or Qwen VL) which converts the output of the vision encoder to embeddings to be fed into the model. These embeddings still don't have to be in a joint embedding space, but if you take the first transformer layer of the decoder and get the key/values of the embeddings, these should be in the same space as they would need to be for self attention to work.

**Comment by dfloreaa** [comment]
Your are completely right, working with an early fusion model would make much more sense for what I am currently trying to achieve. From what I've seen, PaliGemma 2 just came out and may be useful for my usecase. I will look further into the alignment method they use for their embeddings.

Thank you very much! Your intel has been quite helpful, there seems not to be an implementation for either PaliGemma nor QwenVL on the torchtune library, so I may have to look into some alternatives.

Hope you have a great day,
Daniel
